jena-extras/jena-serviceenhancer/pom.xml

```diff
@@ -0,0 +1,185 @@
+    <version>4.6.0-SNAPSHOT</version>
+  </parent>
+
+  <properties>
+    <automatic.module.name>org.apache.jena.serviceplugins</automatic.module.name>
+  </properties>
+
+  <dependencies>
+    <dependency>
+      <groupId>org.apache.jena</groupId>
+      <artifactId>apache-jena-libs</artifactId>
+      <type>pom</type>
+    </dependency>
+
+    <dependency>
+      <groupId>com.github.ben-manes.caffeine</groupId>
+      <artifactId>caffeine</artifactId>
+      <version>3.1.1</version>
+    </dependency>
+
+    <!--
+      <dependency>
+        <groupId>org.aksw.commons</groupId>
+        <artifactId>aksw-commons-io-buffers</artifactId>
+        <version>0.9.5-SNAPSHOT</version>
+      </dependency>
+      <dependency>
+        <groupId>org.aksw.commons</groupId>
+        <artifactId>aksw-commons-claimingcache</artifactId>
+        <version>0.9.5-SNAPSHOT</version>
+      </dependency>
+    -->
+
+    <dependency>
+      <groupId>org.xenei</groupId>
+      <artifactId>junit-contracts</artifactId> 
+      <scope>test</scope>
+      <exclusions>
+        <exclusion>
+          <artifactId>commons-cli</artifactId>
+          <groupId>commons-cli</groupId>
+        </exclusion>
+        <exclusion>
+          <artifactId>commons-logging</artifactId>
+          <groupId>commons-logging</groupId>
+        </exclusion>
+      </exclusions>
+    </dependency>
+
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>jcl-over-slf4j</artifactId>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+    	<groupId>org.mockito</groupId>
+    	<artifactId>mockito-core</artifactId> 
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>org.apache.logging.log4j</groupId>
+      <artifactId>log4j-slf4j-impl</artifactId>
+      <scope>test</scope>
+    </dependency>
+
+  </dependencies>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-jar-plugin</artifactId>
+        <configuration>
+          <archive>
+            <manifestEntries>
+              <Automatic-Module-Name>${automatic.module.name}</Automatic-Module-Name>
+            </manifestEntries>
+          </archive>
+        </configuration>
+      </plugin>
+
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-source-plugin</artifactId>
+        <executions>
+          <execution>
+            <id>attach-sources</id>
+            <!-- <phase>package</phase> package is the default -->
+            <goals>
+              <goal>jar-no-fork</goal>
+            </goals>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
+
+  <!-- Profile to build a "plugin jar bundle" that can be loaded with fuseki or any other jena-based app -->
```

Comment:
There are better ways to do that in Fuseki/Main.

It's actually easier to separate building the extensions and packaging as a FusekiModule into two maven modules. Exclusion management then by <excludes> on artifacts.

It's certainly easier to test the FusekiModule.

Reply:
I am not totally sure what setup you prefer here - the poms of which modules should be adjusted and how?

The current functionality of the serviceenhancer extension works with fuseki without the need for a fuseki module - the service loader of the serviceenhancer registers an assembler that can be used to construct a dataset that can be exposed via a fuseki service.
That means that just adding the serviceenhancer plugin as a dependency to fuseki directly would so far be sufficient to enable configuration via assembler.

There is another issue however which I am not sure how to tackle best: Execution of the `<java:org.apache.jena.sparql.service.enhancer.function.cacheInvalidate>` function should be restricted to only authorized users. Otherwise malicious agents could disrupt a service by constantly invalidation all caches.

Reply:
The difference between a FusekiModule and a JenaSystem is that FusekiModules are drop-ins. The base Fuseki server build does not have a dependency on it. (That said - I see why a JenaSystem should not work as a drop-in as well for now.)

Pragmatically, we can't keep loading more into Fuseki - some dependencies are huge (GraalVM, for example).

Drop-ins are added to the classpath by the server script. I've push up a _draft_ fuseki script:

  https://github.com/afs/jena/blob/fuseki-main-script/jena-fuseki2/jena-fuseki-server/fuseki-main

which picks up `lib/*`.

Fuseki/webapp already has FUSEKI_CLASSES but that's a bit harder for non-Java people.

Reply:
> `<java:org.apache.jena.sparql.service.enhancer.function.cacheInvalidate>` function should be restricted to only authorized users

This isn't a good use of functions! Functions get moved around by the optimizer.

The way to do it is to separate Fuseki server endpoint for cache control with its own security setup.

Reply:
Hm, it's actually quite common to have access control on functions (or database objects in general).
E.g. postgres has `GRANT ALL ON PROCEDURE <procedure_name> TO <role>;` and Virtuoso has something similar such as `grant execute on DB.DBA.SPARQL_DELETE_DICT_CONTENT to "SPARQL";`

With the functions one can e.g. filter out which cache entries to invalidate:
```sparql
PREFIX sepf: <java:org.apache.jena.sparql.service.enhancer.pfunction.>
PREFIX sefn: <java:org.apache.jena.sparql.service.enhancer.function.>

SELECT sefn:cacheInvalidate(?id) {
  ?id sepf:cacheList (?serviceIri ?queryStr ?binding)
  FILTER(CONSTAINS(?queryStr, '?s'))
}
```
Yes, functions can get moved around by the optimizer but that's more an argument that one has to be aware of pitfalls rather than that the whole approach is flawed. At least, other DBMS have support for ACLs on functions.

And ideally I'd prefer if this kind of ACL could be done on the dataset/context level in order to have it decoupled from the HTTP framework's security. But  retrofitting ACLs on sparql functions is another can of worms - maybe eventually it needs to be opened.

If you are suggesting having cache control outside of the realm of sparql function extensions it don't think it make things easier / streamlined.

Reply:
ARQ does not have the SQL security model.

SQL procedures aren't queries and aren't functions - they are any SQL statements, they have their defined execution semantics as to order and they can have side effects. 

This particular case could be done by context+function registry and/or datasets with different setups over the same data. But that only goes so far.

Reply:
**About security**: I was exactly thinking about having some factory method that creates appropriate graph views from a base dataset, a security model and a given user (actor). This is pretty much what is being done in fuseki with DataAccessCtl and SecurityContext. Well, there is not an explicit DatasetGraph view for a certain agent, but the SecurityContext corresponds to an agent's permissions and there is `SecurityContext.createQueryExecution(Query, DatasetGraph)` to evaluate a query on a dataset w.r.t. a set of permissions.

For security over functions and property functions it seems that what would be needed is an extended or new SecurityContext with additional allow/deny lists and a permission check at an appropriate place (probably in ARQ) upon function evaluation.

> But that only goes so far.
Hm, is there much more complexity to it?
So I think for the first release of _this_ PR the security of the cache management is not that critical but its likely to be eventually needed in practice.

> This particular case could be done by context+function registry and/or datasets with different setups over the same data.
Yes, that was what I was aiming for.

**W.r.t to fuseki modules:** We are already making use of https://github.com/afs/jena/blob/fuseki-main-script/jena-fuseki2/apache-jena-fuseki/fuseki-server (rather than https://github.com/afs/jena/blob/fuseki-main-script/jena-fuseki2/jena-fuseki-server/fuseki-main) which loads all jars from the `./extra` folder. Not sure why there are two scripts with seemingly similar purpose - is the latter becoming the new start script?

The criticized shade profile in the serviceenhancer module is there to create the the self-contained jar that can be placed into the `extras` folder and thus on fuseki's classpath. So I think the main question here is which jena maven module should I use an example setup?

Reply:
> Not sure why there are two scripts with seemingly similar purpose

Different Fuseki's : webapp and main.

Reply:
> About security:

Why does the service execution need additional security? This is all becoming a separate processing system. Without careful whole system consideration, there will be leaks (e.g. the context) that reveal information.

Reply:
The security / ACL is relevant for the aforementioned feature of listing the cache content using a property function and for explicitly invalidating cache entries using a function:

```sparql
SELECT sefn:cacheInvalidate(?id) {
  ?id sepf:cacheList (<http://dbpedia.org/sparql> ?queryStr ?binding)
  FILTER(CONSTAINS(?queryStr, '?s'))
}
```

The same principle could be applied to e.g. list and cancel (long) running queries with SPARQL such as via a yasgui akin to  postgresql's `SELECT pg_cancel_backend(pid) FROM pg_stat_activity WHERE state = 'active' and pid <> pg_backend_pid();`.
The issue here is that these should actually be privileged functions not everyone is permitted to run.

And yes, if there were additional SPARQL extension functions that allowed for arbitrary reads and modification of the context then of course it could reveal sensitive information.

Reply:
> is the latter becoming the new start script?

Script `apache-jena-fuseki/fuseki-server` runs `jena-fuseki-fulljar` (not `jena-fuseki-server`) renamed into `apache-jena-fuseki` as `fuseki-server.jar`.

The goal is replace `jena-fuseki-fulljar`, a webapp system, in favour of Fuseki/Main together with modules for the UI and admin functionality.

`jena-fuseki-war` remains but it will be the only webapp based form.

Teh webapp version does not support FusekiModules. Only Fuseki/Main has that extension mechanism.

Reply:
> The issue here is that these should actually be privileged functions not everyone is permitted to run.

That is bundling everything into SPARQL queries. This is not SQL.

Privileged operations do not need to go through SPARQL query or be restricted to SPARQL syntax.

One step necessary step is ban <java:>.

```
PREFIX sepf: <java:org.apache.jena.sparql.service.enhancer.pfunction.>
PREFIX sefn: <java:org.apache.jena.sparql.service.enhancer.function.>
```
These need registered URIs.

e.g. create a separate endpoint API that can be protected.

> The same principle could be applied to e.g. list and cancel (long) running queries
Which is what the tasks endpoints can be extended for.

Reply:
Same namespace prefix as the assembler (see another comment). From a whole system POV, whether you want to split function/property function/assembler vocab or use the same namespace is a local style choice. Either works.

`/extra/` isn't necessary. (would we have "/service-enhancer" as well? No.)

Reply:
The plugin now only introduces the namespace `http://jena.apache.org/service-enhancer#` which is used for both ARQ context symbols and assembler configuration.